nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py

```diff
@@ -103,24 +105,33 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
             _speech_head_embedding.weight.data.fill_(0)
             _speech_head_embedding.shared = True
             list_of_speech_tokens_embeddings.append(_speech_head_embedding)
-            list_of_speech_heads.append(MegatronTokenLevelHead(_speech_head_embedding.weight.size(0), False))
+            if speech_head_type == 'token_level':
+                list_of_speech_heads.append(MegatronTokenLevelHead(_speech_head_embedding.weight.size(0), False))
+            elif speech_head_type == 'linear':
+                # Linear layer that maps from hidden size to speech codebook size
+                hidden_size = self.frozen_model.enc_dec_model.decoder_cfg.hidden_size
+                list_of_speech_heads.append(torch.nn.Linear(hidden_size, speech_codebook_size))
 
         self.frozen_model.enc_dec_model.speech_tokens_heads = torch.nn.ModuleList(list_of_speech_heads)
         self.frozen_model.enc_dec_model.speech_tokens_embeddings = torch.nn.ModuleList(
             list_of_speech_tokens_embeddings
         )
 
-        # TODO: remove hardcoding
-        self.frozen_model.enc_dec_model.speech_residual_model_1 = SimplestModule(
-            self.frozen_model.enc_dec_model.decoder_cfg.hidden_size, speech_offset + speech_codebook_size
-        )
-        self.frozen_model.enc_dec_model.speech_residual_model_2 = SimplestModule(
-            self.frozen_model.enc_dec_model.decoder_cfg.hidden_size, speech_codebook_size
-        )
+        if speech_head_type == 'token_level':
```

Comment:
if speech_head_type == linear, do we not need speech_residual_models ?

Reply:
So for linear head we dont need them. Linear head simply means map the decoder transformer's output to the vocab scores using a linear layer - This is what was suggested by the Musicgen paper so I tried that. 

The other option still uses those speech_residual_models but not the linear head.

Reply:
I see I missed the `elif self.speech_head_type == "linear"` condition in `token_level_encoder_decoder.py` file. Makes sense.